### PR TITLE
Connect the fluentd forwarder to the Fluentd server through TLS without user and password

### PR DESCRIPTION
**Pull request by Skeptor**
|Related issue|
|---|
|https://github.com/wazuh/wazuh/issues/3900|

## Description

The fluent forwarder was requiring a username and password to complete the handshake with the fluent server. This isn't necessary when the forwarder is connected through TLS.

This PR sets the username and the password to an empty string when the variables are NULL and remove the restriction of requiring them to do the handshake.

## Logs/Alerts example

When the username and the password are empty the agent will log a debug message:
```
2019/09/03 13:50:58 fluent-forward[19385] wm_fluent.c:599 at wm_fluent_handshake(): DEBUG: Connection with localhost:24224 established
2019/09/03 13:50:58 fluent-forward[19385] wm_fluent.c:610 at wm_fluent_handshake(): DEBUG: Fluent server requires user name and password, but they are undefined.
2019/09/03 13:50:58 fluent-forward[19385] wm_fluent.c:633 at wm_fluent_handshake(): INFO: Connected to host 'demo' (localhost:24224)
```

## Tests

- Compilation without warnings in every supported platform
  - [x] Linux
  - [ ] Windows
  - [ ] MAC OS X
- [x] Source installation
- [ ] Package installation
- [ ] Source upgrade
- [ ] Package upgrade
- [x] Review logs syntax and correct language
- [ ] QA templates contemplate the added capabilities

- Memory tests for Linux
  - [x] Scan-build report
  - [ ] Coverity
  - [x] Valgrind (memcheck and descriptor leaks check)
  - [ ] Dr. Memory
  - [ ] AddressSanitizer
- Memory tests for Windows
  - [ ] Scan-build report
  - [ ] Coverity
  - [ ] Dr. Memory
- Memory tests for macOS
  - [ ] Scan-build report
  - [ ] Leaks
  - [ ] AddressSanitizer

- [ ] Retrocompatibility with older Wazuh versions
- [ ] Working on cluster environments
- [ ] Configuration on demand reports new parameters
- [ ] The data flow works as expected (agent-manager-api-app)
- [ ] Added unit tests (for new features)
- [ ] Stress test for affected components

### CODE REVIEW

**Comment by Skeptor** [comment]
Scan build
---------
```
make[1]: se sale del directorio '/mnt/HDD/wazuh/src'

Done building server

scan-build: Removing directory '/tmp/scan-build-2019-09-03-120706-7130-1' because it contains no reports.
scan-build: No bugs found.
```

**Comment by Skeptor** [comment]
Make test
----------

```
root@pablo-work:/mnt/HDD/wazuh/src# make test TARGET=server
make DEBUG=1 TEST=1 tap_os_crypto tap_os_net tap_os_regex tap_shared tap_os_zlib tap_os_xml tap_fluentd_forwarder
make[1]: se entra en el directorio '/mnt/HDD/wazuh/src'
make[1]: 'tap_os_crypto' está actualizado.
make[1]: 'tap_os_net' está actualizado.
make[1]: 'tap_os_regex' está actualizado.
make[1]: 'tap_shared' está actualizado.
make[1]: 'tap_os_zlib' está actualizado.
make[1]: 'tap_os_xml' está actualizado.
make[1]: 'tap_fluentd_forwarder' está actualizado.
make[1]: se sale del directorio '/mnt/HDD/wazuh/src'
make run_tests
make[1]: se entra en el directorio '/mnt/HDD/wazuh/src'


   STARTING TEST - OS_CRYPTO   

    ok  1 - Blowfish encryption test.
    ok  2 - MD5 encryption test.
    ok  3 - MD5 file reading encryption test.
    ok  4 - MD5 non-existing file to read from.
    ok  5 - SHA1 encryption test.
    ok  6 - SHA1 file reading encryption test.
    ok  7 - SHA1 non-existing file to read from.
    ok  8 - MD5+SHA1 file reading encryption test.
    ok  9 - MD5+SHA1 reading from file using command encryption test.
    ok 10 - MD5+SHA1 non-existing file to read from using command.
1..10

      [ ALL TESTS PASSED ]

    ENDING TEST  - OS_CRYPTO   



    STARTING TEST - OS_NET   

    ok  1 - TCP IPV4 send an receive: localhost test.
    ok  2 - TCP IPV4 send an receive: inet test.
    ok  3 - TCP IPV6 send an receive test.
    ok  4 - TCP invalid socket test.
    ok  5 - UDP IPV4 send an receive test.
    ok  6 - UDP IPV6 send an receive test.
    ok  7 - UDP invalid socket test.
    ok  8 - Unix sockets test.
    ok  9 - Unix invalid socket test.
    ok 10 - Get host by name: google dns test.
    ok 11 - Get host by name: null host test.
    ok 12 - Get host by name: non-existent host test.
1..12

      [ ALL TESTS PASSED ]

    ENDING TEST  - OS_NET   



    STARTING TEST - OS_REGEX   

    ok  1 - Matching strings with patterns using OS_Match2 test.
    ok  2 - Not matching strings with patterns using OS_Regex test.
    ok  3 - Matching strings with patterns using OS_Regex test.
    ok  4 - Not matching strings with patterns using OS_Regex test.
    ok  5 - Matching strings with patterns using OS_WordMatch test.
    ok  6 - Not matching strings with patterns using OS_WordMatch test.
    ok  7 - Matching strings with patterns using OS_StrIsNum test.
    ok  8 - Not matching strings with patterns using OS_StrIsNum test.
    ok  9 - Matching strings with patterns using OS_StrHowClosedMatch test.
    ok 10 - OS_StrBreak test.
    ok 11 - OSRegex_Execute test.
    ok 12 - Validate the characters of a hostname.
    ok 13 - Validate case insensitive char map test.
    ok 14 - Validate back slash char test.
    ok 15 - Validate charset from '\d'.
    ok 16 - Validate charset from '\w' test.
    ok 17 - Validate charset from '\s' test.
    ok 18 - Validate charset from '\p' test.
    ok 19 - Validate charset from '\D' test.
    ok 20 - Validate charset from '\W' test.
    ok 21 - Validate charset from '\S' test.
    ok 22 - Validate charset from '\.' test.
    ok 23 - Validate charset from '\t' test.
    ok 24 - Validate left parenthesis '(' char test.
    ok 25 - Validate right parenthesis ')' char test.
    ok 26 - Validate dollar '$' char test.
    ok 27 - Validate OR '|' char test.
    ok 28 - Validate less than '<' char test.
    ok 29 - Validate string starts with substring test.
    ok 30 - Not matching substring at the beginning of string.
    ok 31 - There is no race condition in OSRegex_Execute_ex().
1..31

      [ ALL TESTS PASSED ]

    ENDING TEST  - OS_REGEX   



   STARTING TEST - OS_SHARED   

    ok  1 - Search and replace strings test.
    ok  2 - Filter a random string into UTF-8 with character replacement.
    ok  3 - Filter a random string into UTF-8 without character replacement.
1..3

      [ ALL TESTS PASSED ]

   ENDING TEST  - OS_SHARED   



   STARTING TEST - OS_ZLIB   

    ok  1 - Compress and uncompress a regular string test.
    ok  2 - Compress and uncompress a regular string with '\n', '\t' and '\r' test.
    ok  3 - Try to compress using ((void *)0) as source.
    ok  4 - Try to compress using ((void *)0) as destination.
    ok  5 - Try to compress using 0 as destination size.
    ok  6 - Try to uncompress using ((void *)0) as source.
    ok  7 - Try to uncompress using ((void *)0) as destination.
    ok  8 - Try to uncompress using 0 as source size.
    ok  9 - Try to uncompress using 0 as destination size.
1..9

      [ ALL TESTS PASSED ]

   ENDING TEST  - OS_ZLIB   



   STARTING TEST - OS_XML   

    ok  1 - Simple XML node test.
    ok  2 - Multiple XML nodes test.
    ok  3 - Childen on XML nodes test.
    ok  4 - Multiple content on XML node test.
    ok  5 - Attributes inside XML test.
    ok  6 - Variables inside XML test.
    ok  7 - Comments inside XML test.
    ok  8 - Special chars inside XML test.
    ok  9 - Line counter inside XML test.
    ok 10 - Invalid XML test.
    ok 11 - Unclosed node inside XML test.
    ok 12 - Unclosed second node inside XML test.
    ok 13 - Unclosed comment inside XML test.
    ok 14 - Node not opened inside XML test.
    ok 15 - Unclosed attribute inside XML test.
    ok 16 - Unquoted attribute inside XML test.
    ok 17 - Infinite attribute inside XML test.
    ok 18 - Invalid variable name inside XML test.
    ok 19 - Invalid variable name (2*MAX_SIZE) inside XML test.
    ok 20 - Invalid variable name (2*MAX_SIZE) inside XML test.
    ok 21 - Empty attribute inside XML test.
    ok 22 - Empty attribute start inside XML test.
    ok 23 - Invalid attribute closing inside XML test.
    ok 24 - Wrong variable name inside XML test.
    ok 25 - Unknown variable inside XML test.
    ok 26 - Invalid attribute closing (no space) inside XML test.
    ok 27 - Invalid attribute closing (no second quote) inside XML test.
    ok 28 - Duplicated attribute inside XML test.
    ok 29 - No attribute value inside XML test.
    ok 30 - No first attribute value inside XML test.
    ok 31 - OS_RootElementExist test.
    ok 32 - OS_GetOneContentforElement test.
    ok 33 - OS_WriteXML test.
    ok 34 - OS_WriteXML failures test.
    ok 35 - OS_GetAttributeContent test.
    ok 36 - OS_GetContents test.
    ok 37 - OS_GetElementContent test.
    ok 38 - OS_GetElements test.
    ok 39 - OS_GetAttributes test.
    ok 40 - Node name inside XML overflow test.
    ok 41 - Node value inside XML overflow test.
    ok 42 - Node attribute name inside XML overflow test.
    ok 43 - Node attribute value inside XML overflow test.
1..43

      [ ALL TESTS PASSED ]

   ENDING TEST  - OS_XML   



   STARTING TEST - FLUENTD FORWARDER MODULE  

2019/09/03 16:43:53 fluent-forward: ERROR: No tag defined.
    ok  1 - Test configuration no 'tag' defined.
2019/09/03 16:43:53 fluent-forward: ERROR: No socket_path defined.
    ok  2 - Test configuration no 'socket_path' defined.
2019/09/03 16:43:53 fluent-forward: INFO: No client address defined. Using localhost.
    ok  3 - Test configuration no 'address' defined.
2019/09/03 16:43:53 fluent-forward: ERROR: Invalid timeout value (negative)
    ok  4 - Test configuration invalid 'timeout' defined.
2019/09/03 16:43:53 fluent-forward: WARNING: No shared_key defined. SSL is disabled and the user_name option won't apply.
2019/09/03 16:43:53 fluent-forward: ERROR: Password required because user_name is defined
    ok  5 - Test configuration no 'password' defined.
    ok  6 - Test configuration valid configuration
    ok  7 - Test connection
2019/09/03 16:43:53 fluent-forward: INFO: Connected to host localhost:24224
    ok  8 - Test handshake
2019/09/03 16:43:53 fluent-forward: INFO: Connected to host localhost:24224
    ok  9 - Test send message
    ok 10 - Test configuration dump
1..10

      [ ALL TESTS PASSED ]

   ENDING TEST  - FLUENTD FORWARDER MODULE   

make[1]: se sale del directorio '/mnt/HDD/wazuh/src'
```

File: src/wazuh_modules/wm_fluent.c [hunk 1]
```diff
@@ -607,9 +607,8 @@ static int wm_fluent_handshake(wm_fluent_t * fluent) {
         }
 
         if (helo->auth_size > 0) {
```
Comment (JuantAldea) [inline src/wazuh_modules/wm_fluent.c]: According to the protocol, receiving an auth message does not imply that you have to provide a non-empty user/password and it should not alert if they are empty.
Moreover, given that now credentials are always filled, this debug message will always trigger.

File: src/wazuh_modules/wm_fluent.c [hunk 2]
```diff
@@ -607,9 +607,8 @@ static int wm_fluent_handshake(wm_fluent_t * fluent) {
...
-            if (!(fluent->user_name && fluent->user_pass)) {
-                mwarn("Authentication error: the Fluent server requires user name and password.");
-                goto end;
+            if (!strcmp(fluent->user_name, "") && !strcmp(fluent->user_pass, "")) {
+                mdebug1("Fluent server requires user name and password, but they are undefined.");
             }
         } else if (fluent->user_name || fluent->user_pass) {
```
Comment (JuantAldea) [inline src/wazuh_modules/wm_fluent.c]: Same as before here. Given that now credentials are always filled, this debug message will always trigger.

**Comment by snaow** [comment]
Ping @JuantAldea

File: src/config/wmodules-fluent.c [hunk 3]
```diff
@@ -195,6 +193,14 @@ int wm_fluent_read(xml_node **nodes, wmodule *module)
         }
     }
 
+    if (!fluent->user_name) {
+        os_strdup("",fluent->user_name);
```
Comment (JuantAldea) [inline src/config/wmodules-fluent.c]: Fix spacing.

**Comment by JuantAldea** [comment]
> After testing, there seem to be a strange behaviour when the server is configured without "authentication yes" but the client sends some credentials, making our client hangs in the PONG step.
> 
> ```
> #2  0x00007ffff7b26a7e in sock_read () from /var/ossec/bin/../lib/libwazuhext.so
> #3  0x00007ffff7b21043 in BIO_read () from /var/ossec/bin/../lib/libwazuhext.so
> #4  0x00007ffff7aaea9a in ssl3_read_n () from /var/ossec/bin/../lib/libwazuhext.so
> #5  0x00007ffff7ab33a3 in ssl3_get_record () from /var/ossec/bin/../lib/libwazuhext.so
> #6  0x00007ffff7ab0435 in ssl3_read_bytes () from /var/ossec/bin/../lib/libwazuhext.so
> #7  0x00007ffff7ab7535 in ssl3_read () from /var/ossec/bin/../lib/libwazuhext.so
> #8  0x00007ffff7ac2c89 in SSL_read () from /var/ossec/bin/../lib/libwazuhext.so
> #9  0x00005555555e8500 in wm_fluent_recv (fluent=0x555555692620, unp=0x7ffff56f1de0) at wazuh_modules/wm_fluent.c:304
> #10 0x00005555555e85d7 in wm_fluent_unpack (fluent=0x555555692620, unp=0x7ffff56f1de0, result=0x7ffff56f1dc0)
>     at wazuh_modules/wm_fluent.c:318
> #11 0x00005555555e9507 in wm_fluent_recv_pong (fluent=0x555555692620) at wazuh_modules/wm_fluent.c:535
> #12 0x00005555555e99cb in wm_fluent_hs_tls (fluent=0x555555692620) at wazuh_modules/wm_fluent.c:601
> #13 0x00005555555e9b11 in wm_fluent_handshake (fluent=0x555555692620) at wazuh_modules/wm_fluent.c:631
> #14 0x00005555555e79c8 in wm_fluent_main (fluent=0x555555692620) at wazuh_modules/wm_fluent.c:99
> #15 0x00007ffff7956182 in start_thread (arg=<optimized out>) at pthread_create.c:486
> #16 0x00007ffff787fb1f in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
> ```
> 
> When it comes to the server side, the log is
> 
> ```
> 2019-09-25 17:05:13 +0200 [trace]: #0 fluent/log.rb:281:trace: accept tcp connection (ssl session not established yet)
> 2019-09-25 17:05:13 +0200 [debug]: #0 fluent/log.rb:302:debug: starting server
> 2019-09-25 17:05:13 +0200 [trace]: #0 fluent/log.rb:281:trace: session instances: all=1 closed=0
> 2019-09-25 17:05:13 +0200 [trace]: #0 fluent/log.rb:281:trace: accepting ssl session
> 2019-09-25 17:05:13 +0200 [debug]: #0 fluent/log.rb:302:debug: generating helo
> ```

After hacking around a bit, it seems we are not honoring the protocol, as the PING message requires the password or an empty string, but the later is not being added.

A question raises as well: 
In the case when authentication is not required but username and passwords are specified in our configuration, what should we do?

**Pull request merged**